velox/exec/tests/RowContainerTest.cpp

```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
+    return numNulls;
+  }
+
+  // Given a list of values sorted in ascending order with nulls first re-orders
+  // the elements to reverse the order and / or pu nulls last.
+  template <typename T>
+  void changeSortingOrder(
+      std::vector<std::optional<T>>& values,
+      bool ascending,
+      bool nullsFirst) {
+    if (ascending) {
       if (nullsFirst) {
-        std::reverse(expectedOrder.begin() + 1, expectedOrder.end());
+        // Nothing to do. 'values' are expected to be already sorted in
+        // ascending order with nulls first.
       } else {
-        std::reverse(expectedOrder.begin(), expectedOrder.end());
+        // Move the nulls from the start of the list to the end of the list.
+        const auto numNulls = countLeadingNulls(values);
+        for (auto i = 0; i < values.size() - numNulls; ++i) {
+          values[i] = values[i + numNulls];
+        }
+
+        for (auto i = values.size() - numNulls; i < values.size(); ++i) {
+          values[i].reset();
+        }
       }
     } else {
-      if (!nullsFirst) {
-        for (int i = 0; i < expectedOrder.size() - 1; ++i) {
-          expectedOrder[i] = expectedOrder[i + 1];
-        }
-        expectedOrder[expectedOrder.size() - 1] = std::nullopt;
+      if (nullsFirst) {
+        // Keeps nulls at the start of the list. Reverse the order of remaining
+        // non-null elements.
+        const auto numNulls = countLeadingNulls(values);
+        std::reverse(values.begin() + numNulls, values.end());
+      } else {
+        // Reverse the order of all elements.
+        std::reverse(values.begin(), values.end());
       }
     }
-    VectorPtr expected = vectorMaker.flatVectorNullable<T>(expectedOrder);
+  }
+
+  // A list of rows with unique zero-based indices.
+  using IndexedRows = std::vector<std::pair<vector_size_t, char*>>;
+
+  VectorPtr copyValues(
+      const IndexedRows& indexedRows,
+      const VectorPtr& values) {
+    const auto numRows = indexedRows.size();
+    auto copy = BaseVector::create(values->type(), numRows, pool_.get());
+    for (auto i = 0; i < numRows; ++i) {
+      copy->copy(values.get(), i, indexedRows[i].first, 1);
+    }
+    return copy;
+  }
+
+  static std::vector<char*> storeRows(
+      DecodedVector& decoded,
+      vector_size_t numRows,
+      RowContainer& rowContainer) {
+    std::vector<char*> rows(numRows);
+    for (size_t i = 0; i < numRows; ++i) {
+      rows[i] = rowContainer.newRow();
+      rowContainer.store(decoded, i, rows[i], 0);
+    }
+
+    return rows;
+  }
+
+  static IndexedRows indexRows(const std::vector<char*>& rows) {
+    IndexedRows indexedRows;
+    indexedRows.reserve(rows.size());
+    for (auto i = 0; i < rows.size(); ++i) {
+      indexedRows.push_back({i, rows[i]});
+    }
+    return indexedRows;
+  }
+
+  template <typename T>
+  void testRowContainerCompareAPIs(
+      const TypePtr& type,
+      const VectorPtr& values,
+      const VectorPtr& expected,
+      bool ascending,
```

Comment:
maybe pass std::optional<CompareFlags>; this will make the code a bit easier to read

Reply:
If I'm understanding correctly, then it means we would skip passing in ascending/nullsFirst for the default values? Not sure what the purpose of using optional would be.

Reply:
With std::optional<CompareFlags> compareFlags argument we would pass compareFlags to the API based on whether the flags are set or not. I think this will make code cleaner and more readable.